### PR TITLE
[4.2] Smart Search: Index custom fields

### PR DESCRIPTION
**Pull request by Hackwar**
Pull Request for Issue #12829.

### Summary of Changes
This PR implements indexing of the custom fields for a component. This is controlled on a per field basis, where you can select if the content should be just added to the search index or handled as a taxonomy or both or actually none of the above. You can then implement this feature for your respective finder plugin with a single line of code.
Each custom field has a new option in its edit screen where you can select how the indexer should treat this field. The finder plugin then has to call `Helper::addCustomFields($item, 'context');` and the rest is just magic. ;)
Right now this is implemented for com_content and com_contacts.

### Testing Instructions
1. Your installation should have some custom fields for com_content or com_contact in it.
2. Edit a custom field and set the "Smart Search" option to "Make searchable and add as taxonomy".
3. Run the indexing process either by running the mass indexer in the smart search component or by editing an item and saving it again.
4. See that the custom field shows up as a taxonomy and can be find by searching for its value.

### Documentation Changes Required
The documentation needs to be amended to contain how to implement this feature for third party finder plugins.

### Notes
1. Since 4.1 is in RC, this will have to go into 4.2. Since there is no 4.2-dev branch yet, I'm pointing this towards 4.1-dev for now.

### CODE REVIEW

File: administrator/components/com_fields/forms/field.xml [hunk 1]
```diff
@@ -327,6 +327,20 @@
...
+						type="list"
+						label="COM_FIELDS_FIELD_SEARCHINDEX_LABEL"
+						default="0"
+						validate="options"
+				>
+					<option value="0">COM_FIELDS_FIELD_SEARCHINDEX_DONT</option>
+					<option value="1">COM_FIELDS_FIELD_SEARCHINDEX_SEARCHABLE</option>
+					<option value="2">COM_FIELDS_FIELD_SEARCHINDEX_TAXONOMY</option>
+					<option value="3">COM_FIELDS_FIELD_SEARCHINDEX_BOTH</option>
+				</field>
```
Comment (brianteeman) [inline administrator/components/com_fields/forms/field.xml]:
```suggestion
				<field
					name="searchindex"
					type="list"
					label="COM_FIELDS_FIELD_SEARCHINDEX_LABEL"
					default="0"
					validate="options"
					>
					<option value="0">COM_FIELDS_FIELD_SEARCHINDEX_DONT</option>
					<option value="1">COM_FIELDS_FIELD_SEARCHINDEX_SEARCHABLE</option>
					<option value="2">COM_FIELDS_FIELD_SEARCHINDEX_TAXONOMY</option>
					<option value="3">COM_FIELDS_FIELD_SEARCHINDEX_BOTH</option>
				</field>
```

File: administrator/components/com_finder/src/Indexer/Helper.php [hunk 2]
```diff
@@ -112,7 +113,7 @@ public static function tokenize($input, $lang, $phrase = false)
 		$tokens = array();
 		$terms = $language->tokenise($input);
 
-		// @todo: array_filter removes any number 0's from the terms. Not sure this is entirely intended
+		// TODO: array_filter removes any number 0's from the terms. Not sure this is entirely intended
```
Comment (brianteeman) [inline administrator/components/com_finder/src/Indexer/Helper.php]: please revert this change

**Comment by Hackwar** [comment]
This also solves #13338

**Comment by laoneo** [comment]
Nice, how would the data be represented on the site part in a search result?

**Comment by Hackwar** [comment]
> Nice, how would the data be represented on the site part in a search result?

If it is a taxonomy, it would simply use the given string in a dropdown. Otherwise it wouldn't display the custom fields by default. For output, you should create your own layout, maybe load custom fields again or similar stuff.

File: administrator/language/en-GB/com_fields.ini [hunk 3]
```diff
@@ -57,8 +57,15 @@ COM_FIELDS_FIELD_RENDER_CLASS_LABEL="Render Class"
 COM_FIELDS_FIELD_RENDEROPTIONS_HEADING="Render Options"
 COM_FIELDS_FIELD_REQUIRED_LABEL="Required"
 COM_FIELDS_FIELD_SAVE_SUCCESS="Field saved"
+COM_FIELDS_FIELD_SEARCHINDEX_BOTH="Make searchable and add as taxonomy"
+COM_FIELDS_FIELD_SEARCHINDEX_DONT="Don't make searchable"
+COM_FIELDS_FIELD_SEARCHINDEX_LABEL="Searchindex"
+COM_FIELDS_FIELD_SEARCHINDEX_SEARCHABLE="Make searchable"
+COM_FIELDS_FIELD_SEARCHINDEX_TAXONOMY="Add as taxonomy"
+
 COM_FIELDS_FIELD_SHOWLABEL_DESC="Show or Hide the label when the field renders."
```
Comment (Quy) [inline administrator/language/en-GB/com_fields.ini]:
```suggestion
COM_FIELDS_FIELD_SEARCHINDEX_TAXONOMY="Add as taxonomy"
COM_FIELDS_FIELD_SHOWLABEL_DESC="Show or Hide the label when the field renders."
```

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on 92b39ea937340d40f0d84030af70ce18158eb61e<br /><br />Tested with three different fields: 
- Text field in Content
- Text field in Contacts
- List field in Contacts
Tried the different parameters in the fields.
It works as described.

Great PR! Thank you Hannes.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36747">issues.joomla.org/tracker/joomla-cms/36747</a>.</sub>

**Comment by Quy** [comment]
I have not checked all field types so should this setting be excluded such as media, list of images...?

**Comment by coolcat-creations** [comment]
We installed this PR with the prebuilt package and smart search does not work at all (even not with normal content)

- Plugins are enabled
- Fields are setup
- Menu item is set up
- Content is successfully indexed



<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36747">issues.joomla.org/tracker/joomla-cms/36747</a>.</sub>

**Comment by Hackwar** [comment]
Can you be a bit more specific? Does it return any search results or do you get a white screen in the frontend or backend? Since this only changes how stuff is indexed and doesn't do anything related to the actual searching, this might even be a new bug.

**Comment by coolcat-creations** [comment]
Article Content is for example cheeseburger and when you search for cheeseburger, the article can't be found at all. It says that there are no search results for that.

**Comment by Hackwar** [comment]
I think I found the issue and it is not related to this PR. I'm guessing that you wrote an article, saved it and then went to the frontend searching for it, right? We are storing the current time and date as publish_start date. In the SQL query to retrieve the results, we do compare that timestamp with the current time minus the seconds. That last part is done so that such a query doesn't change every second and the database has a chance to cache it. So if you wait a few more seconds until the next minute ticks around, the results should show up. I was able to recreate exactly that behavior. But: All of this is not related to this PR.

**Comment by webnet-assmann** [comment]
I have tested this item :white_check_mark: successfully on 92b39ea937340d40f0d84030af70ce18158eb61e<br /><br />testet with 3 different custome fields:
Content Field: tet and editor
contact field: text
Thanks a lot Hannes!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36747">issues.joomla.org/tracker/joomla-cms/36747</a>.</sub>

**Comment by chmst** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36747">issues.joomla.org/tracker/joomla-cms/36747</a>.</sub>

**Comment by heelc29** [comment]
@rdeutz Can you add this to milestone

**Comment by joomdonation** [comment]
@Hackwar Look like we attempted to index custom fields before already in fields system plugin https://github.com/joomla/joomla-cms/blob/4.2-dev/plugins/system/fields/fields.php#L512

However, look like that code is not working at all. With the feature introduce in this PR, what should we do with that existing code ?

**Comment by Hackwar** [comment]
That code is pretty much useless. We should delete it entirely.

**Comment by roland-d** [comment]
@Hackwar I have reverted this PR in light of the issues found in issue #38222 Please review the findings and propose a new PR which we can use. Thank you.
